Go faster

g-kimbell opened the pull request:
Trying to make this go faster, as we have a lot of large files to parse frequently.

Includes the changes from #8, fixing #4. Also fixes #3. 

Main logic:
- Use np.frombuffer to parse the binary data
  - Read whole file in one go, operate on slices, avoid any copying
  - Avoids having to do Python while/for loops  
- Wait until as late as possible to convert uts to Timestamp objects
  - Faster to do interpolation etc. on floats 
- Wait until as late as possible to convert Status from int to the string codes
  - Speeds up cycle number generation, can operate faster on ints and avoids a string split in a loop
 - Switch from pandas to polars, compute as much as possible in parallel
 - Unzip and process the ndc files in parallel

Final result is around 10x speed up, particularly on large files with ~>100k rows. For smaller files the changes are less dramatic.

![image](https://github.com/user-attachments/assets/c28261f5-b142-40cb-b637-f95b6c5dcebb)

g-kimbell commented:
Polars is incredibly fast, we can get reading 820k from 7.3 s -> 280 ms with CPU utilization 9%->56%, it takes another ~100 ms to convert to pandas and convert all the types after that. I'm working on getting it running for all filetypes.

d-cogswell commented:
Thanks for investigating this, I haven't spent much effort on performance so there's probably room for gains. In my own work I primarily use NewareNDA for preprocessing files to either Arrow or Parquet, and then I run analysis on the preprocessed data.

g-kimbell commented:
This seems to be working for me now for all files I can test, with a speed up of around 10-15x. Regression tests fail because we are now reading some extra rows and interpolating differently, there are some floating point errors, and the categorical columns behave slightly differently (using enum instead of categorial).

@d-cogswell @Grimler91 Do either of you have ndc2 or ndc5 files? I cannot find any to test changes.

g-kimbell commented:
Great, thank you for the additional files! Agreed on other fixes first, and I still need to iron out some details which the new data will help a lot with.

g-kimbell commented:
Added polars dependency, got rid of tzlocal as we can do it the same way as before with pandas and datetime. I got some very exotic crashes trying to do timezones in polars then converting to pandas.

Hopefully I have time to look at ndc2 and 5 in the next week.

Grimler91 commented:
Thanks, gitattributes sounds like a good idea! :)

Tried it out, had to install pyarrow or else: 
```
    df = df.to_pandas()
  File "/lib/python3.13/site-packages/polars/dataframe/frame.py", line 2549, in to_pandas
    return self._to_pandas_without_object_columns(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        self, use_pyarrow_extension_array=use_pyarrow_extension_array, **kwargs
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/lib/python3.13/site-packages/polars/dataframe/frame.py", line 2600, in _to_pandas_without_object_columns
    record_batches = df._df.to_pandas()
ModuleNotFoundError: No module named 'pyarrow'
```

so probably it should be added to dependencies as well. 

I got a ~2x increase, but I am then also doing some postprocessing that can likely be optimised further..

Grimler91 commented:
There is some issue with read_ndc_11_filetype_5, I'm getting the wrong aux V values. Will try to investigate during the day.

Grimler91 commented:
@g-kimbell sent you an email!

g-kimbell commented:
Thanks, I forgot to remove the 0.0 V rows after parsing those files - it was looking for "Voltage" but the aux columns are "V"

Still, both this PR and the development branch don't get 100% the same values as BTSDA, they are correct to 4 decimal places ~99.4% of the time, but something is still wrong. I'll keep looking into it.

g-kimbell commented:
This now uses the masks on the first byte == 101, like the original

Grimler91 commented:
Last commit changed line endings again, so it is hard to see what exactly you changed (in github UI at least) 🙈 

<img width="661" height="440" alt="line_ending" src="https://github.com/user-attachments/assets/8cc81c75-7774-4911-a25b-8cb9843e50a8" />

g-kimbell commented:
Sorry, since #18 .gitattributes will force all line endings to LF now. After this pain it should be consistent

`_read_ndc` now has a `mask` argument, if you provide that and give it a column named 'mask' it filters on that

`_read_ndc_11_filetype_5` takes the first byte as the mask value, and filters where it == 101

This is a vectorised version of what happens in the development branch
```
for i in struct.iter_unpack('<cfh', bytes[132:-2]):
    if i[0] == b'\x65':
        # add row
```

Grimler91 commented on NewareNDA/NewareNDAx.py:
```diff
@@ -1,609 +1,588 @@
+    with open(file, "rb") as f, mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ) as mm:
+        # Get ndc file version and filetype
+        [ndc_filetype] = struct.unpack("<B", mm[0:1])
+        [ndc_version] = struct.unpack("<B", mm[2:3])
```
Would you mind re-adding the debug log here? I think that one has been useful to quickly see what version a file has

```
logger.debug(f"NDC version: {ndc_version} filetype: {ndc_filetype}")
```

g-kimbell commented:
I added the new sorting method, and fixed silly mistakes with interpolating capacity and energy.

I've tested it against 22 files of various types that I have access to vs BTSDA. I can only compare filetype 17 to the development because my BTSDA somehow crashes when opening it. My tests are all working now, except for a couple of issues:

There are a couple of remaining problems, but maybe we can solve them in future PRs:

- The file in #3 returns one column as `T99` instead of `T1`, I don't know how we can fix this when `TestInfo.xml` says `AuxID: 99`.
- Neware are a bit indecisive whether Step is the monotonic step number or the step index, Dan raised this in #15. Our step is always monotonically increasing but BTSDA can return it both ways.
- I noticed that `123456789012_Unit27_Example_Data_File` has duplicate indexes, both in this and the development branch, BTSDA doesn't.
- `123456789012_Unit27_Example_Data_File` also has a weird tiny offset on aux temperature, they agree if you do round(1), temperature only has 0.1'C precision anyway.

d-cogswell commented:
Hi @g-kimbell thanks for your hard work here. This PR is going to be difficult to review since it's almost a complete rewrite of the code, and unfortunately all my regression tests are failing for various reasons. Can you please break your work into several smaller PRs:
-A PR to fix #3
-A separate PR to address interpolation issues
-A PR for speed improvements without changing existing dependencies
-A separate PR for speed improvements with polars

The speed improvements might be tricky, and shouldn't hold up fixing the other issues. If it involves rewriting large sections of code, it also might make sense to do this in a fork of the project.

g-kimbell commented:
Okay I'll try to break this down into new PRs and do #3 and #4 first. For the speed improvements, changing from `iterunpack` to `readfrombuffer` and the change from pandas to polars will be a big rewrite, but the logic is unchanged.
